src/index/hnsw/faiss_hnsw.cc

```diff
@@ -0,0 +1,1398 @@
+        return knowhere::IndexEnum::INDEX_FAISS_HNSW_PRQ;
+    }
+
+ protected:
+    std::unique_ptr<faiss::IndexProductResidualQuantizer> tmp_index_prq;
+
+    Status
+    TrainInternal(const DataSetPtr dataset, const Config& cfg) override {
+        // number of rows
+        auto rows = dataset->GetRows();
+        // dimensionality of the data
+        auto dim = dataset->GetDim();
+
+        // config
+        auto hnsw_cfg = static_cast<const FaissHnswPrqConfig&>(cfg);
+
+        auto metric = Str2FaissMetricType(hnsw_cfg.metric_type.value());
+        if (!metric.has_value()) {
+            LOG_KNOWHERE_ERROR_ << "Invalid metric type: " << hnsw_cfg.metric_type.value();
+            return Status::invalid_metric_type;
+        }
+
+        // create an index
+        const bool is_cosine = IsMetricType(hnsw_cfg.metric_type.value(), metric::COSINE);
+
+        // HNSW + PRQ index yields BAD recall somewhy.
+        // Let's build HNSW+FLAT index, then replace FLAT with PRQ
+
+        std::unique_ptr<faiss::IndexHNSW> hnsw_index;
+        if (is_cosine) {
+            hnsw_index = std::make_unique<faiss::IndexHNSWFlatCosine>(dim, hnsw_cfg.M.value());
+        } else {
+            hnsw_index = std::make_unique<faiss::IndexHNSWFlat>(dim, hnsw_cfg.M.value(), metric.value());
+        }
+
+        hnsw_index->hnsw.efConstruction = hnsw_cfg.efConstruction.value();
+
+        // prq
+        faiss::AdditiveQuantizer::Search_type_t prq_search_type =
+            (metric.value() == faiss::MetricType::METRIC_INNER_PRODUCT)
+                ? faiss::AdditiveQuantizer::Search_type_t::ST_LUT_nonorm
+                : faiss::AdditiveQuantizer::Search_type_t::ST_norm_float;
+
+        std::unique_ptr<faiss::IndexProductResidualQuantizer> prq_index;
+        if (is_cosine) {
+            prq_index = std::make_unique<faiss::IndexProductResidualQuantizerCosine>(
+                dim, hnsw_cfg.m.value(), hnsw_cfg.nrq.value(), hnsw_cfg.nbits.value(), prq_search_type);
+        } else {
+            prq_index = std::make_unique<faiss::IndexProductResidualQuantizer>(
+                dim, hnsw_cfg.m.value(), hnsw_cfg.nrq.value(), hnsw_cfg.nbits.value(), metric.value(), prq_search_type);
+        }
+
+        // should refine be used?
+        std::unique_ptr<faiss::Index> final_index;
+        if (hnsw_cfg.refine.value_or(false)) {
+            // yes
+            auto final_index_cnd = pick_refine_index(data_format, hnsw_cfg.refine_type, std::move(hnsw_index));
+            if (!final_index_cnd.has_value()) {
+                return Status::invalid_args;
+            }
+
+            // assign
+            final_index = std::move(final_index_cnd.value());
+        } else {
+            // no refine
+
+            // assign
+            final_index = std::move(hnsw_index);
+        }
+
+        // we have to convert the data to float, unfortunately, which costs extra RAM
+        auto float_ds_ptr = convert_ds_to_float(dataset, data_format);
```

Comment:
Does FAISS have a plan to support FP/BF16?

Reply:
We can ask Matthijs about supporting fp16 / bf16, but I doubt that he will go for it. So, I'm using a workaround for now.